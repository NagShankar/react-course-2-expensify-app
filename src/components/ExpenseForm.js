import React from "react";
import moment from "moment";
import "react-dates/initialize";
import { SingleDatePicker } from "react-dates";
//import "react-dates/lib/css/_datepicker.css"; //moved to app.js so other components can make use of it, just to avoid confusion of inporting it everywhere

//const now = moment();
//console.log(now.format('MMM Do, YYYY'))

class ExpenseForm extends React.Component {
  constructor(props) {
    super(props); //now props will have the 'expense prop' sent from Edit Expense page to populate the fields to edit it
    this.state = {
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
      amount: props.expense ? props.expense.amount.toString() : "", //converting back to string while editing which we had earlier converted to number before submitting it to store
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(), //providing initial value i.e by default today's date
      calendarFocused: false,
      error: "",
    };
  }

  //WE CAN COMBINE DESCRIPTION AND NOTE IN ONE EVENT HANDLER LIKE BELOW
  //onValueChange = (e) => {
  //    const name=e.target.name;
  //    const value=e.target.value;
  //
  //    this.setState({
  //        [name]:value
  //    })
  //
  //}

  //   componentDidMount() {
  //     console.log(this.state);
  //   }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  onNoteChange = (e) => {
    const note = e.target.value;
    //you cant assign e.target.value in setState directly, you cant do like this, this will throw an synthetic error, if you really want to assign target value directly in setState use e.persist() before doing that as suggested by that error
    //e.persist();
    //this.setState(()=>({note:e.target.value}))
    this.setState(() => ({ note }));
  };
  onAmountChange = (e) => {
    const amount = e.target.value;
    //below matches empty string and regex with it
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };
  onDateChange = (date) => {
    //, it activates a calendar again, it takes an argument(can be Date or Now and any name) in the form of moment and gives the back the date to reset the date
    //not letting user to clear date field, now setting onChange only when there's date already present there, thus forcing date to be present all the time and not letting user to chnage it at all, and change only if there's date
    //console.log(date.valueOf()); //new moment object is returned, and valueOf will convert time and date in milliseconds since Unix Epoch
    if (date) {
      this.setState(() => ({ createdAt: date }));
    }
  };
  onFocusChange = ({ focused }) => {
    //as given in official docs, destructuring the focused property, which have boolean value = TRUE on clicking its
    //console.log(focused);
    this.setState(() => ({ calendarFocused: focused }));
  };

  onSubmit = (e) => {
    e.preventDefault(); //avoiding refresh if they press submit, which is browser default behaviour

    //validation to check if the value is present for fields like description and amount, only then make submit
    if (!this.state.description || !this.state.amount) {
      //alert("please enter value for desc and amount")
      //set error state to "Please provide description and amount"
      this.setState(() => ({
        error: "Please provide description and amount",
      }));
    } else {

      this.props.onSubmission({
        description: this.state.description,
        note: this.state.note,
        amount: parseFloat(this.state.amount, 10),
        createdAt: this.state.createdAt.valueOf(), //if valueOf not provided, then it displays moment object
      });
      //console.log('submitted');
        
        this.setState(() => ({
        //now emptying all fields after success
        description: "",
        note: "",
        amount: "",
        createdAt: moment(), //providing initial value
        calendarFocused: false,
        error: "",
      }));
        
        
        
    }
  };

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            name="description"
            placeholder="Description"
            value={this.state.description}
            onChange={this.onDescriptionChange}
            autoFocus
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />

          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={
              1
            } /*change this number to see as many calendars you want */
            isOutsideRange={() =>
              false
            } /*enables past/finished dates as well, you can set range also, refer doc */
          />

          <textarea
            name="note"
            placeholder="Add a note for your expense(optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          ></textarea>
          <button>Add Expense</button>
        </form>
      </div>
    );
  }
}

export default ExpenseForm;
