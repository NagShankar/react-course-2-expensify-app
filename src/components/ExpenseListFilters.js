import React from "react";
import { connect } from "react-redux";
import { DateRangePicker } from "react-dates";
import {
  setTextFilter,
  sortByAmount,
  sortByDate,
  setStartDate,
  setEndDate,
} from "./../actions/filters";

//converted to class component
export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null,
  };

  //onDatesChange take an object with destructured start date and end date, refer doc
  changingDate = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  //just for changing the focus, no real impact on this project
  changeFocus = (calendarFocused) => {
    //console.log(calendarFocused); //either start date or end date
    this.setState(() => ({ calendarFocused: calendarFocused }));
  };


  onTextChange = (e) => {
            this.props.setTextFilter(e.target.value);
          }

  onSortChange = (e) => {
            //alert(e.target.value)
            if (e.target.value === "amount") {
              this.props.sortByAmount();
            } else if (e.target.value === "date") {
              this.props.sortByDate();
            }
    };

  render() {
    return (
      <div>
        <input
          type="text"
          value={this.props.filters.text}
          onChange={this.onTextChange}
        />

        <select
          value={this.props.filters.sortBy}
          onChange={this.onSortChange}>
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>

        <DateRangePicker
          startDateId={"dwjkhqkehwqjkeq"}
          endDateId={"cxzvcxbzbczxbz"}
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.changingDate}
          focusedInput={this.state.calendarFocused} //this is mandatory, with initial value null, has no impact for this project
          onFocusChange={this.changeFocus} //this is mandatory, will focus if its Start date and End date by making use of calendarFocused, has no impact for this project
          showClearDates={true} //clear the date range by clicking on "X" mark
          numberOfMonths={
            1
          } /*change this number to see as many calendars you want */
          isOutsideRange={() =>
            false
          } /*enables past/finished dates as well, you can set range also, refer doc */
        />
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  //accessing state of the store
  return {
    filters: state.filters, //grabbing only filters state
  };
};

const mapDispatchToProps = (dispatch) => ({
   //using same name as disptach names
    
    setTextFilter : (text) =>  dispatch(setTextFilter(text)),
    sortByDate : () => dispatch(sortByDate()),
    sortByAmount : () => dispatch(sortByAmount()),
    setStartDate : (startDate) => dispatch(setStartDate(startDate)),
    setEndDate : (endDate) => dispatch(setEndDate(endDate))
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
