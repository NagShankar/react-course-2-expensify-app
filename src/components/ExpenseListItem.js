import React from "react";
//import { connect } from 'react-redux';
//import { removeExpense } from './../actions/expenses'
import { Link } from 'react-router-dom';

//now destructuring instead od taking props
const ExpenseListItem = ({id, description, amount, createdAt}) => {
  return(
    <div>
          <Link to={`/edit/${id}`}><h3>{description} </h3></Link>
          <h4>Amount: {amount} Created: {createdAt}</h4>
         
          {/* lets have remove button only when wants to edit an expense*/} 
         
          {/* 
              <button onClick={()=>{
                    dispatch(removeExpense({id:id})); //remove expense accepts "id" as an object, check actions/expenses.js for more details, using dispatch after destructuring which was available inside props as props.dispatch
               }}>Remove</button>

           */} 
      </div>
  );    
}

//export default connect()(ExpenseListItem); //CONNECTING GIVE DISPATCH AS A PROP

export default ExpenseListItem;