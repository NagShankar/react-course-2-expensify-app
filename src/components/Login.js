import React from "react";
import { connect } from "react-redux";
import { startLogin } from "./../actions/auth";

export const Login = ({ startLogin }) => { //if you import this then you need curly braces in file which you're importing because its named export
    return(
      <div> 
         <h1>Welcome To Expensify App</h1>
          <h5>Please Login Below</h5>
           <button onClick={startLogin}>Login</button> 
        </div>
    
    )
}

//OR... if you want to use mapDispatchToProps
//const mapDispatchToProps = (dispatch) => {
//    return {
//        startLogin: () => dispatch(startLogin())
//    }
//}

export default connect(null,{ startLogin })(Login); //no need of curly braces for this while importing