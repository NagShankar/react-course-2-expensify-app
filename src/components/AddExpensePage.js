import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
//...........after adding middleware and modifying addExpense action creator
//import { addExpense } from "./../actions/expenses";
import { startAddExpense } from "./../actions/expenses";


//previosly this was functional component, now changed to class for testing the unconnected version
export class AddExpensePage extends React.Component{
    
    onSubmission = (expense) => {
        //console.log(expense);
        //props.dispatch(addExpense(expense));
         this.props.submitNow(expense);
         this.props.history.push(
          "/"
        ); /* routing back to dashboard page after submission using history given by react router */
      }
    
    render(){
        
        return(
         <div>
               <h1>Add Expense</h1>
               <ExpenseForm onSubmission={this.onSubmission} />
         </div>
        
        );
    }
    
    
}


const mapDispatchToProps = (dispatch) => {
    return{
        submitNow : (expense) => dispatch(startAddExpense(expense)) //changed addExpense(expense) to startAddExpense(expense), i.e we are dispatching async call here
   }
}

export default connect(undefined, mapDispatchToProps)(AddExpensePage); //connect gives access to props which allow us to dispatch action
