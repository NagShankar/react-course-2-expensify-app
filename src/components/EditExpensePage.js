import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { editExpense, removeExpense } from "./../actions/expenses";

//previosly this was functional component, now changed to class for testing the unconnected version
export class EditExpensePage extends React.Component{
    
    onSubmission = (expense) => {
          //console.log(expense)
        //alert("hi")
          this.props.editExpense(this.props.expense.id, expense);
          this.props.history.push("/");
        }
    removeNow = (e) => {
         //alert("remove now")
          this.props.removeNow({ id: this.props.expense.id }); //remove expense action accepts "id" as an object, check actions/expenses.js for more details
          this.props.history.push("/"); //redirecting after deleting
        }
       
    render(){
        
         //console.log(this.props); //now props will have the matched expense inside props
        return(
                        
            <div>
           Editing Expense for id: {this.props.match.params.id}
                <ExpenseForm 
                      expense={this.props.expense} /* passing the matched expense as a prop to ExpenseForm component which then can be used to set the default and populate the fields with that data */
                      onSubmission={this.onSubmission}
                />
               <button onClick={this.removeNow}> Remove </button>
          </div>
            
        );
    }
    
}


const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) => {
      return expense.id === props.match.params.id; //returning the matched expense object from the expenses array
    }),
  };
};

const mapDispatchToProps = (dispatch, props) => ({
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  removeNow: (data) => dispatch(removeExpense( data ))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
