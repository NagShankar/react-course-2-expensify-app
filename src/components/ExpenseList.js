import React from "react";
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import getVisibleFilters from './../selectors/expenses'

export const ExpenseList = (props) => {
    return(
     <div className="content-container">
        <div className="list-header">
           <div className="show-for-mobile">Expenses</div>
           <div className="show-for-desktop">Expense</div>
           <div className="show-for-desktop">Amount</div>
        </div>
        
      
       {/* {props.expenses.length}
        {props.filters.text} */}
            <div className="list-body">
                    {
                        props.expenses.length === 0 ? (<div className="list-item list-item--message"><span>No expenses!</span></div>) : (props.expenses.map((expense)=>{
                          return <ExpenseListItem key={expense.id} {...expense} /> 
                       }))

                    }
            </div>
          
        {/* spreading the remaining data as ...expense */}

        </div>
    
    );
}

//const ConnectedExpenseList = connect((state)=>{ //first argument is the state from the store, can be given any name
//    return{
//      //name:"nag"
//        expenses:state.expenses
//    };
//})(ExpenseList);
//
//export default ConnectedExpenseList;


//instead of manually assigning to variable, exporting and then importing it, we can have unnamed export like below, this is most prefered way

//instead passing function to Connect as first argument, its common practice to store that function in widely used variable "mapStateToProps" it literally means we are mapping store's State to Props to use it inside component

//when store chnages, the state here(in argument) gets automatically updated and stays upto date with the current data and then re-renders the component when soething chnages, so no need for component to subscribe
const mapStateToProps = (state)=> { //first argument is the state from the store, can be given any name
    return{
      //name:"nag"
        expenses:getVisibleFilters(state.expenses, state.filters),
        //filters:state.filters //we dont want to take data directly from state, because its not filtered or sorted, we want it filtered every time hence sending it to getVisibleFilters component
    };
};

export default connect(mapStateToProps)(ExpenseList);

 



//using it in dashboard file