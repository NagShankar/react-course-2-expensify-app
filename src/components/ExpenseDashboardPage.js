import React from "react";
import ExpenseList from './ExpenseList'; //importing connected version of the component, so no need of "{}"
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummary from './ExpensesSummary';

const ExpenseDashboardPage = () => (
 <div>
       <ExpensesSummary />
       <ExpenseListFilters />
       <ExpenseList />
           
    </div>
);


export default ExpenseDashboardPage;