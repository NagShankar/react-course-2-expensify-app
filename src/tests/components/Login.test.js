import React from "react";
import { Login } from '../../components/Login'; //grabbing named export, we want unconnected version
import { shallow } from 'enzyme';

test('should render Login page', ()=>{
    const wrapper=shallow(<Login startLogin={ () => { } }/>);
    expect(wrapper).toMatchSnapshot();                     
})

test('should call startlogin on button click', () => {
    const onStartloginCLick = jest.fn();
    const wrapper = shallow(<Login startLogin={onStartloginCLick}/>); 
    wrapper.find('button').simulate('click');
    expect(onStartloginCLick).toHaveBeenCalled() //fails if the fake function(onStartlogoutCLick) is not called
})
