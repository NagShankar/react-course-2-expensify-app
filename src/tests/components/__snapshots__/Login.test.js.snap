// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Login page 1`] = `
<div>
  <h1>
    Welcome To Expensify App
  </h1>
  <h5>
    Please Login Below
  </h5>
  <button
    onClick={[Function]}
  >
    Login
  </button>
</div>
`;
